README.md
# Q Applet: Mini Meters

Displays the CPU usage, free memory, GPU usage, or average load on individual keys on a Das Keyboard Q Series.

[GitHub repository](https://github.com/corykim/daskeyboard-applet--mini-meters)

## Example

Three meters enabled on three separate keys.

![Mini Meters on a Das Keyboard Q](assets/image.png "Mini Meters applet")

## Changelog

[CHANGELOG.MD](CHANGELOG.md)

## Installation

Requires a Das Keyboard Q series: www.daskeyboard.com

Installation, configuration and uninstallation of applets is done within
the Q Desktop application (https://www.daskeyboard.com/q)

## Running tests

- `yarn test`

## Acknowledgements

- Application Icon made by Freepik from www.flaticon.com
- GPU meter code written by [Brandon Schabell](https://github.com/brandonschabell)

## Contributions

Pull requests welcome.

## Copyright / License

Copyright 2019 Cory Kim

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

# [1.2.0](https://github.com/corykim/daskeyboard-applet--mini-meters/compare/v1.1.1...v1.2.0) (2019-02-21)


### Bug Fixes

* **Tests:** Expect `[#000000](https://github.com/corykim/daskeyboard-applet--mini-meters/issues/000000)` instead of literal `black` ([f53f9de](https://github.com/corykim/daskeyboard-applet--mini-meters/commit/f53f9de))


### Features

* **initial release:** Update collateral ([8463fb6](https://github.com/corykim/daskeyboard-applet--mini-meters/commit/8463fb6))
* **nVidia GPU:** Allow test to gracefully fail if no nvidia GPU present ([6d2fcb6](https://github.com/corykim/daskeyboard-applet--mini-meters/commit/6d2fcb6))
* **nVidia GPU:** Remove redundant `async` notation ([2f478b6](https://github.com/corykim/daskeyboard-applet--mini-meters/commit/2f478b6))
* **nVidia GPU Meter:** Add nVidia GPU meter from Brnadon Schabell ([a7fca7a](https://github.com/corykim/daskeyboard-applet--mini-meters/commit/a7fca7a))
* **Signal:** Refactor to single-key signal ([8b959dc](https://github.com/corykim/daskeyboard-applet--mini-meters/commit/8b959dc))



#### 1.0.0 (2019-02-20)
Initial version.


test/test.js
const assert = require('assert');
const t = require('../index');

const black = '#000000';

describe('MiniMeter', function () {
  describe('#getMetric', function () {
    it('#getMetric(default)', async function () {
      return buildApp().then(app => app.getMetric(t.ModeEnum.CPU).then(percent => {
        console.log("CPU Usage is: " + percent);
        assert.ok(percent);
      }))
    });
    it('#getMetric(freemem)', async function () {
      return buildApp().then(app => app.getMetric(t.ModeEnum.FREE_MEM).then(percent => {
        console.log("Free Mem is: " + percent);
        assert.ok(percent);
      }))
    });
    it('#getMetric(average)', async function () {
      return buildApp().then(app => app.getMetric(t.ModeEnum.AVERAGE_LOAD).then(percent => {
        console.log("Average Load is: " + percent);
        // Windows always return 0
        assert.ok(percent || percent === 0);
      }))
    });
    it('#getMetric(nvidiaGpu)', async function () {
      return buildApp().then(app => app.getMetric(t.ModeEnum.NVIDIA_GPU).then(percent => {
        console.log("nVidia GPU is: " + percent);
        // Does not work unless nVidia GPU is present?
        assert.ok(percent || percent === 0);
      }).catch(error => {
        console.error("Is an nVidia GPU present?");
      }))
    });
  })

  it('#getColor', async function () {
    return buildApp().then(app => {
      assert.equal(black, app.getColor(99), 'Expected big number to return black');
      assert.equal(black, app.getColor(-1), 'Expected negative number to return black');

      assert(app.getColor(0), 'Expected 0 to return a color');
      assert.notEqual(black, app.getColor(0), 'Expected 0 to return a color');

      assert(app.getColor(.999), 'Expected .999 to return a color');
      assert.notEqual(black, app.getColor(.999), 'Expected .999 to return a color');
    })
  });

  it('#generatePoints', async function () {
    return buildApp().then(app => {
      for (let i = 0; i < 1; i += 0.01) {
        const points = app.generatePoints(i);
        assert.equal(1, points.length)
        assert.notEqual('black', points[0].color);
        assert(points[0].color.match(/^#[A-Fa-f0-9]{6}$/));
      }
    })
  });

  it('#run()', async function () {
    return buildApp().then(async app => {
      try {
        const signal = await app.run();
        console.log(JSON.stringify(signal));
        assert.ok(signal);
        assert.ok(signal.points[0][0], "I should have at least one point");
      }
      catch (error) {
        assert.fail(error);
      }
    })
  });

  it('deletes old signals', function () {
    return buildApp().then(app => {
      return app.deleteOldSignals();
    })
  })
});

async function buildApp(config) {
  config = config || {};
  let app = new t.MiniMeter();
  app.config = {
    mode: t.ModeEnum.CPU,
    geometry: {
      height: 1,
      width: 1,
      origin: {
        x: 1,
        y: 1
      }
    },

    ...config
  };

  return app;
}

track-cpu.js
// Library to make track cpuUsage
// if not installed run npm install request
var os = require('os-utils');
// Library to make simplified HTTP client requests
// if not installed run npm install request
var request = require('request');


var backendUrl = 'http://localhost:27301';
var headers = {
  "Content-Type": "application/json"
}


// using vector to target the second line of the DK5Q keyboard
const zoneIds = ['1,1', '2,1', '3,1', '4,1', '5,1', '6,1', '7,1', '8,1', '9,1', '10,1'];
// Color associated to each zone
const colors = ['#00FF00', '#00FF00', '#00FF00', '#00FF00', '#FFFF00', '#FFFF00', '#FF0000', '#FF0000', '#FF0000', '#FF0000'];



/** post a signal to the localAPI  */
function sendSignal(signal) {
  // HTTP POST request to the cloud
  request.post({
    url: backendUrl + '/api/1.0/signals',
    headers: headers,
    body: signal,
    json: true
  });
}

/** get a color of a zone depending on it's index on the zone array */
function getColor(zoneIndex) {
  return colors[zoneIndex];
}

function getCpuUsage() {
  os.cpuUsage(function (v) {
    const numberOfKeys = zoneIds.length;
    // multiply the cpu percentage by the number total of keys 
    const numberOfKeysToLight = Math.round(numberOfKeys * v) + 1;
    zoneIds.forEach((zone, index) => {
      // if the zone is after the number max of keys to light. Turn off the zone light
      if (index > numberOfKeysToLight) {
        // Construct the signal to send
        var signal = {
          'zoneId': zone.toString(),
          'color': '#000000',
          'effect': 'SET_COLOR',
          'pid': 'DK5QPID',
          'isMuted': 'true',
          'clientName': 'Node script',
          'message': 'cpeUsage',
          'name': 'cpeUsage'
        };
        sendSignal(signal);
      } else {
        // turn on the zone with the proper color
        var signal = {
          'zoneId': zone.toString(),
          'color': getColor(index, numberOfKeysToLight),
          'effect': 'SET_COLOR',
          'pid': 'DK5QPID',
          'isMuted': 'true',
          'clientName': 'Node script',
          'message': 'cpuUsage',
          'name': 'cpuUsage'
        };
        sendSignal(signal);
      }
    });


  });
}

setInterval(() => {
  getCpuUsage();
}, 2000);





q-cpu-usage.js
// Library to track cpuUsage
const os = require('os-utils');

// Library to send signal to Q keyboards
const q = require('daskeyboard-applet');

// Color associated to cpu activity from low (green) to high (red).
const colors = ['#00FF00', '#00FF00', '#00FF00', '#00FF00', '#FFFF00', '#FFFF00', '#FF0000',
  '#FF0000', '#FF0000', '#FF0000'
];

const logger = q.logger;


class CpuUsage extends q.DesktopApp {
  constructor() {
    super();
    // run every 3000 ms
    this.pollingInterval = 3000;
    logger.info("CPU Usage Meter ready to go!");
  }

  // call this function every pollingInterval
  async run() {
    return this.getCpuUsage().then(percent => {
      return new q.Signal({
        points: [this.generatePoints(percent)],
        name: "CPU Usage",
        message: Math.round(percent * 100) + "%",
        isMuted: true, // don't flash the Q button on each signal
      });
    });
  }

  async getCpuUsage() {
    return new Promise((resolve) => {
      os.cpuUsage(v => {
        resolve(v);
      })
    })
  }

  generatePoints(percent) {
    const numberOfKeys = 10;
    // multiply the cpu percentage by the number total of keys 
    const numberOfKeysToLight = Math.round(numberOfKeys * percent);
    let points = [];

    // create a list of points (zones) with a color). Each point 
    // correspond to an LED
    for (let i = 0; i < numberOfKeys; i++) {
      points.push(new q.Point(this.getColor(i, numberOfKeysToLight)));
    }

    return points;
  }

  /** get a color of a zone depending on it's index on the zone array */
  getColor(zoneIndex, numberOfKeysToLight) {
    if (zoneIndex >= numberOfKeysToLight) {
      // if the zone is after the number max of keys to light. Turn off the light
      // Black color = no light
      return '#000000';
    } else {
      // turn on the zone with the proper color
      return colors[zoneIndex];
    }
  }
}

module.exports = {
  CpuUsage: CpuUsage
};

const cpuUsage = new CpuUsage();


index.js
// Library to track cpuUsage
const os = require('os-utils');

// Library to send signal to Q keyboards
const q = require('daskeyboard-applet');

// Library to process colors
const colorsys = require('colorsys');

// Library to track gpuUsage
const smi = require('node-nvidia-smi');

const logger = q.logger;

const ModeEnum = Object.freeze({
  CPU: 'cpu',
  FREE_MEM: 'freemem',
  AVERAGE_LOAD: 'avg1',
  NVIDIA_GPU: 'nvidiaGpu',
});

// the default mode, if none is specified
const ModeDefault = ModeEnum.CPU;

// defines the properties of each available mode
const ModeMap = {
  cpu: {
    name: 'CPU Usage',
    metric: function () {
      return new Promise((resolve) => {
        os.cpuUsage(v => {
          resolve(v);
        })
      })
    }
  },
  avg1: {
    name: 'Average Load',
    metric: async function () {
      return os.loadavg(1);
    }
  },
  freemem: {
    name: 'Free Memory',
    metric: async function () {
      return os.freememPercentage();
    }
  },
  nvidiaGpu: {
    /*
      Credit to Brandon Schabell for GPU Usage applet
      https://github.com/brandonschabell/daskeyboard-applet-gpu-usage
    */
    name: 'GPU',
    metric: function () {
      return new Promise((resolve, reject) => {
        smi(function (err, data) {
          // handle errors
          if (err) {
            logger.error(err);
            reject(err);
          } else {
            var memory_object = data.nvidia_smi_log.gpu.fb_memory_usage
            var used = parseInt(memory_object.used.replace(' MiB', ''))
            var total = parseInt(memory_object.total.replace(' MiB', ''))
            var percent = used / total
            resolve(percent)
            }
        });
      })
    }
  }

}

class MiniMeter extends q.DesktopApp {
  constructor() {
    super();
    // run every 3000 ms
    this.pollingInterval = 3000;
  }

  // call this function every pollingInterval
  async run() {
    let mode = this.config.mode || ModeDefault
    return this.getMetric(mode).then(percent => {
      logger.debug(`Got ${percent} percent`);
      this.deleteOldSignals();

      return new q.Signal({
        points: [this.generatePoints(percent)],
        name: ModeMap[mode].name,
        message: Math.round(percent * 100) + "%",
        isMuted: true, // don't flash the Q button on each signal
      });
    });
  }

  async getMetric(mode) {
    logger.debug(`Mode is ${mode}`);
    if (!mode) {
      mode = ModeDefault;
      logger.debug(`Defaulted mode to ${mode}`);
    }

    return ModeMap[mode].metric();
  }

  /**
   * Delete all previous signals
   */
  async deleteOldSignals() {
    // delete the previous signals
    while (this.signalLog && this.signalLog.length) {
      const signal = this.signalLog.pop().signal;
      logger.debug(`Deleting previous signal: ${signal.id}`)
      await q.Signal.delete(signal).catch(error => {
        logger.error(`Error deleting signal ${signal.id}: ${error}`);
      });

      logger.debug(`Deleted the signal: ${signal.id}`);
    }
  }

  generatePoints(percent) {
    return [new q.Point(this.getColor(percent))];
  }

  getColor(percent) {
    if (percent < 0 || percent > 1) {
      return '#000000' // Black
    } else {
      return colorsys.hslToHex({
        // hue ranges from blue to red
        h: 256 - Math.round(percent * 256),
        s: 100,
        // lightness ranges from 25 - 50
        l: percent * 25 + 25
      });
    }
  }

  async shutdown() {
    await this.deleteOldSignals();
    await super.shutdown();
  }
}

module.exports = {
  MiniMeter: MiniMeter,
  ModeEnum: ModeEnum
};

const miniMeter = new MiniMeter();